File: cli/src/limen/conduct/resources.py
"""Canonical resource ordering and overlap rules for peer-safe parallelism."""

from __future__ import annotations

import posixpath
import re
from dataclasses import dataclass
from pathlib import PurePosixPath

from limen.conduct.models import ResourceClaimV1


_PR_RE = re.compile(
    r"^pr/(?P<owner>[^/]+)/(?P<repo>[^/]+)/(?P<number>[0-9]+)"
    r"/(?P<kind>write|review/(?P<provider>[^@/]+))@(?P<head>[A-Za-z0-9._+-]+)$"
)
_REPO_KIND_RE = re.compile(
    r"^(?P<kind>branch|base|repo-common-dir|agy-scratch|repo)/"
    r"(?P<owner>[^/]+)/(?P<repo>[^/]+)(?:/(?P<rest>.*))?$"
)
_PATH_RE = re.compile(r"^path/(?P<owner>[^/]+)/(?P<repo>[^/]+)/(?P<base>[^/]+)/(?P<prefix>.*)$")


@dataclass(frozen=True)
class Resource:
    raw: str
    kind: str
    repo: str | None = None
    identity: tuple[str, ...] = ()
    prefix: str | None = None


def normalize_key(key: str) -> str:
    key = key.strip().rstrip("/")
    if key.startswith("worktree/"):
        raw_path = key.removeprefix("worktree/") or "/"
        return f"worktree/{posixpath.normpath(raw_path)}"
    pr_match = _PR_RE.fullmatch(key)
    if pr_match:
        kind = (
            f"review/{pr_match.group('provider')}"
            if pr_match.group("provider")
            else "write"
        )
        return (
            f"pr/{pr_match.group('owner').lower()}/{pr_match.group('repo').lower()}/"
            f"{pr_match.group('number')}/{kind}@{pr_match.group('head')}"
        )
    if key.startswith("path/"):
        match = _PATH_RE.fullmatch(key)
        if match:
            prefix = posixpath.normpath("/" + match.group("prefix")).lstrip("/")
            return (
                f"path/{match.group('owner').lower()}/{match.group('repo').lower()}/"
                f"{match.group('base')}/{prefix}"
            ).rstrip("/")
    repo_match = _REPO_KIND_RE.fullmatch(key)
    if repo_match:
        rest = repo_match.group("rest")
        suffix = f"/{rest}" if rest else ""
        return (
            f"{repo_match.group('kind')}/{repo_match.group('owner').lower()}/"
            f"{repo_match.group('repo').lower()}{suffix}"
        )
    return key


def parse_resource(key: str) -> Resource:
    key = normalize_key(key)
    if key.startswith("task/"):
        return Resource(key, "task", identity=(key.removeprefix("task/"),))
    if key.startswith("external/"):
        return Resource(key, "external", identity=(key.removeprefix("external/"),))
    if key.startswith("worktree/"):
        return Resource(key, "worktree", identity=(key.removeprefix("worktree/"),))
    pr_match = _PR_RE.fullmatch(key)
    if pr_match:
        repo = f"{pr_match.group('owner')}/{pr_match.group('repo')}"
        kind = "pr-review" if pr_match.group("provider") else "pr-write"
        identity = (repo, pr_match.group("number"), pr_match.group("provider") or "")
        return Resource(key, kind, repo=repo, identity=identity)
    path_match = _PATH_RE.fullmatch(key)
    if path_match:
        repo = f"{path_match.group('owner')}/{path_match.group('repo')}"
        return Resource(
            key,
            "path",
            repo=repo,
            identity=(repo, path_match.group("base")),
            prefix=str(PurePosixPath("/" + path_match.group("prefix"))),
        )
    repo_match = _REPO_KIND_RE.fullmatch(key)
    if repo_match:
        repo = f"{repo_match.group('owner')}/{repo_match.group('repo')}"
        kind = repo_match.group("kind")
        rest = repo_match.group("rest") or ""
        if kind == "base" and rest.endswith("/integrate"):
            rest = rest[: -len("/integrate")]
            kind = "base-integrate"
        elif kind == "repo-common-dir" and rest == "plumbing":
            kind = "repo-plumbing"
        elif kind == "repo" and rest == "write":
            kind = "repo-write"
        return Resource(key, kind, repo=repo, identity=(repo, rest))
    return Resource(key, "opaque", identity=(key,))


def _prefixes_overlap(left: str, right: str) -> bool:
    left = str(PurePosixPath(left))
    right = str(PurePosixPath(right))
    return left == right or left.startswith(right.rstrip("/") + "/") or right.startswith(left.rstrip("/") + "/")


def resources_overlap(left: ResourceClaimV1, right: ResourceClaimV1) -> bool:
    """Return whether two claims contend under the conduct protocol."""

    a = parse_resource(left.key)
    b = parse_resource(right.key)
    if a.kind == b.kind == "pr-review" and a.identity == b.identity:
        # One provider receipt per exact PR head, even if a caller mislabeled
        # the read-oriented review lease as shared.
        return True
    shareable = {"path"}
    if left.mode == right.mode == "shared" and a.kind in shareable and b.kind in shareable:
        return False
    if a.raw == b.raw:
        return True
    if a.kind == "repo-write" and (a.repo == "*/*" or a.repo == b.repo):
        return b.kind not in {"pr-review"}
    if b.kind == "repo-write" and (b.repo == "*/*" or b.repo == a.repo):
        return a.kind not in {"pr-review"}
    if a.kind == b.kind == "pr-write":
        return a.identity[:2] == b.identity[:2]
    if "pr-review" in {a.kind, b.kind}:
        # Review keys are deliberately independent from writers and from other providers.
        return a.kind == b.kind and a.identity == b.identity
    if a.kind == b.kind == "path":
        return a.identity == b.identity and bool(a.prefix and b.prefix and _prefixes_overlap(a.prefix, b.prefix))
    if a.kind == b.kind == "branch":
        return a.identity == b.identity
    if {a.kind, b.kind} == {"branch", "path"} and a.repo == b.repo:
        branch = a if a.kind == "branch" else b
        path = b if b.kind == "path" else a
        return bool(branch.identity[1] and path.identity[1] == branch.identity[1])
    if a.kind == b.kind == "worktree":
        return a.identity == b.identity
    if a.kind == b.kind and a.kind in {
        "task",
        "external",
        "repo-plumbing",
        "base-integrate",
        "agy-scratch",
        "opaque",
    }:
        return a.identity == b.identity
    return False


def conflicting_keys(
    requested: tuple[ResourceClaimV1, ...] | list[ResourceClaimV1],
    held: tuple[ResourceClaimV1, ...] | list[ResourceClaimV1],
) -> list[tuple[str, str]]:
    return sorted(
        {
            (normalize_key(left.key), normalize_key(right.key))
            for left in requested
            for right in held
            if resources_overlap(left, right)
        }
    )


def sorted_claims(claims: tuple[ResourceClaimV1, ...] | list[ResourceClaimV1]) -> tuple[ResourceClaimV1, ...]:
    dedup: dict[str, ResourceClaimV1] = {}
    for claim in claims:
        normalized = normalize_key(claim.key)
        current = dedup.get(normalized)
        mode = "exclusive" if claim.mode == "exclusive" or current and current.mode == "exclusive" else "shared"
        dedup[normalized] = ResourceClaimV1(key=normalized, mode=mode)
    return tuple(dedup[key] for key in sorted(dedup))

File: web/worker/src/conduct/resources.js
const PR_RE = /^pr\/([^/]+)\/([^/]+)\/([0-9]+)\/(write|review\/([^@/]+))@([A-Za-z0-9._+-]+)$/;
const REPO_KIND_RE = /^(branch|base|repo-common-dir|agy-scratch|repo)\/([^/]+)\/([^/]+)(?:\/(.*))?$/;
const PATH_RE = /^path\/([^/]+)\/([^/]+)\/([^/]+)(?:\/(.*))?$/;

function normalizePosix(value, { absolute = false } = {}) {
  const parts = String(value).split("/");
  const result = [];
  for (const part of parts) {
    if (!part || part === ".") continue;
    if (part === "..") {
      if (result.length && result.at(-1) !== "..") result.pop();
      else if (!absolute) result.push(part);
    } else {
      result.push(part);
    }
  }
  const joined = result.join("/");
  if (absolute) return `/${joined}` || "/";
  return joined || ".";
}

export function normalizeResourceKey(value) {
  let key = String(value || "").trim().replace(/\/+$/, "");
  if (key.startsWith("worktree/")) {
    const rawPath = key.slice("worktree/".length) || "/";
    return `worktree/${normalizePosix(rawPath, { absolute: rawPath.startsWith("/") })}`;
  }
  const pr = key.match(PR_RE);
  if (pr) {
    const kind = pr[5] ? `review/${pr[5]}` : "write";
    return `pr/${pr[1].toLowerCase()}/${pr[2].toLowerCase()}/${pr[3]}/${kind}@${pr[6]}`;
  }
  if (key.startsWith("path/")) {
    const match = key.match(PATH_RE);
    if (match) {
      const prefix = normalizePosix(`/${match[4] || ""}`, { absolute: true }).replace(/^\/+/, "");
      return `path/${match[1].toLowerCase()}/${match[2].toLowerCase()}/${match[3]}/${prefix}`.replace(/\/+$/, "");
    }
  }
  const repo = key.match(REPO_KIND_RE);
  if (repo) {
    const suffix = repo[4] ? `/${repo[4]}` : "";
    return `${repo[1]}/${repo[2].toLowerCase()}/${repo[3].toLowerCase()}${suffix}`;
  }
  return key;
}

export function parseResource(value) {
  const raw = normalizeResourceKey(value);
  if (raw.startsWith("task/")) return { raw, kind: "task", identity: [raw.slice(5)] };
  if (raw.startsWith("external/")) return { raw, kind: "external", identity: [raw.slice(9)] };
  if (raw.startsWith("worktree/")) return { raw, kind: "worktree", identity: [raw.slice(9)] };
  const pr = raw.match(PR_RE);
  if (pr) {
    const repo = `${pr[1]}/${pr[2]}`;
    const provider = pr[5] || "";
    return {
      raw,
      kind: provider ? "pr-review" : "pr-write",
      repo,
      identity: [repo, pr[3], provider],
    };
  }
  const path = raw.match(PATH_RE);
  if (path) {
    const repo = `${path[1]}/${path[2]}`;
    return {
      raw,
      kind: "path",
      repo,
      identity: [repo, path[3]],
      prefix: normalizePosix(`/${path[4] || ""}`, { absolute: true }),
    };
  }
  const repoMatch = raw.match(REPO_KIND_RE);
  if (repoMatch) {
    const repo = `${repoMatch[2]}/${repoMatch[3]}`;
    let kind = repoMatch[1];
    let rest = repoMatch[4] || "";
    if (kind === "base" && rest.endsWith("/integrate")) {
      rest = rest.slice(0, -"/integrate".length);
      kind = "base-integrate";
    } else if (kind === "repo-common-dir" && rest === "plumbing") {
      kind = "repo-plumbing";
    } else if (kind === "repo" && rest === "write") {
      kind = "repo-write";
    }
    return { raw, kind, repo, identity: [repo, rest] };
  }
  return { raw, kind: "opaque", identity: [raw] };
}

function prefixesOverlap(left, right) {
  const a = normalizePosix(left, { absolute: true });
  const b = normalizePosix(right, { absolute: true });
  return a === b || a.startsWith(`${b.replace(/\/+$/, "")}/`) || b.startsWith(`${a.replace(/\/+$/, "")}/`);
}

function sameIdentity(left, right, length = null) {
  const a = length === null ? left : left.slice(0, length);
  const b = length === null ? right : right.slice(0, length);
  return a.length === b.length && a.every((value, index) => value === b[index]);
}

function repoWideMatches(repo, other) {
  return repo === "*/*" || repo === other;
}

export function resourcesOverlap(left, right) {
  const a = parseResource(left.key);
  const b = parseResource(right.key);
  if (a.kind === "pr-review" && b.kind === "pr-review" && sameIdentity(a.identity, b.identity)) {
    return true;
  }
  if (left.mode === "shared" && right.mode === "shared" && a.kind === "path" && b.kind === "path") {
    return false;
  }
  if (a.raw === b.raw) return true;
  if (a.kind === "repo-write" && repoWideMatches(a.repo, b.repo)) return b.kind !== "pr-review";
  if (b.kind === "repo-write" && repoWideMatches(b.repo, a.repo)) return a.kind !== "pr-review";
  if (a.kind === "pr-write" && b.kind === "pr-write") return sameIdentity(a.identity, b.identity, 2);
  if (a.kind === "pr-review" || b.kind === "pr-review") {
    return a.kind === b.kind && sameIdentity(a.identity, b.identity);
  }
  if (a.kind === "path" && b.kind === "path") {
    return sameIdentity(a.identity, b.identity) && Boolean(a.prefix && b.prefix && prefixesOverlap(a.prefix, b.prefix));
  }
  if (a.kind === "branch" && b.kind === "branch") return sameIdentity(a.identity, b.identity);
  if (new Set([a.kind, b.kind]).size === 2 && [a.kind, b.kind].includes("branch") && [a.kind, b.kind].includes("path") && a.repo === b.repo) {
    const branch = a.kind === "branch" ? a : b;
    const path = a.kind === "path" ? a : b;
    return Boolean(branch.identity[1] && branch.identity[1] === path.identity[1]);
  }
  if (a.kind === "worktree" && b.kind === "worktree") return sameIdentity(a.identity, b.identity);
  if (a.kind === b.kind && ["task", "external", "repo-plumbing", "base-integrate", "agy-scratch", "opaque"].includes(a.kind)) {
    return sameIdentity(a.identity, b.identity);
  }
  return false;
}

export function conflictingKeys(requested, held) {
  const pairs = new Set();
  for (const left of requested) {
    for (const right of held) {
      if (resourcesOverlap(left, right)) {
        pairs.add(JSON.stringify([normalizeResourceKey(left.key), normalizeResourceKey(right.key)]));
      }
    }
  }
  return [...pairs].map((pair) => JSON.parse(pair)).sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)));
}

export function sortedClaims(claims) {
  const deduplicated = new Map();
  for (const claim of claims) {
    const key = normalizeResourceKey(claim.key);
    const current = deduplicated.get(key);
    const mode = claim.mode === "exclusive" || current?.mode === "exclusive" ? "exclusive" : "shared";
    deduplicated.set(key, {
      schema_version: "limen.resource_claim.v1",
      key,
      mode,
    });
  }
  return [...deduplicated.keys()].sort().map((key) => deduplicated.get(key));
}
